File: dart/test/client_test.dart
import 'dart:convert';

import 'package:http/http.dart' as http;
import 'package:http/testing.dart';
import 'package:misarmail/misarmail.dart';
import 'package:test/test.dart';

MisarMailClient _client(http.Client mock, {int maxRetries = 3}) =>
    MisarMailClient(
      apiKey: 'mmk_test',
      baseUrl: 'https://api.misar.io/mail/v1',
      maxRetries: maxRetries,
      httpClient: mock,
    );

void main() {
  group('plan', () {
    test('get() returns the plan, allowances and per-feature usage', () async {
      late http.Request seen;
      final mock = MockClient((req) async {
        seen = req;
        return http.Response(
          jsonEncode({
            'plan': {'slug': 'starter', 'name': 'Starter'},
            'sending': {'emails_per_day': 500, 'emails_per_month': 10000},
            'usage': [
              {'feature': 'emails', 'used': 120, 'limit': 500, 'remaining': 380}
            ],
            'upgrade': null,
          }),
          200,
          headers: {'content-type': 'application/json'},
        );
      });

      final out = await _client(mock).plan.get();

      expect(seen.url.path, endsWith('/plan'));
      expect(seen.headers['Authorization'], equals('Bearer mmk_test'));
      expect(out['plan']['slug'], equals('starter'));
      expect(out['usage'][0]['used'], equals(120));
      expect(out['upgrade'], isNull);
    });
  });

  group('plan limits', () {
    // A spent allowance answers 429 — identical by status to a rate limit — so
    // the client must tell them apart by body and not retry the refusal.
    test('a plan refusal is typed and not retried', () async {
      var attempts = 0;
      final mock = MockClient((req) async {
        attempts++;
        return http.Response(
          jsonEncode({
            'error': 'Monthly send allowance spent.',
            'code': 'plan_limit_exceeded',
            'upgrade': {'feature': 'emails', 'currentPlanSlug': 'starter'},
          }),
          429,
          headers: {
            'content-type': 'application/json',
            'x-misar-plan': 'starter',
            'x-misar-upgrade-url': 'https://misarmail.com/pricing',
            'retry-after': '600',
          },
        );
      });

      await expectLater(
        _client(mock).plan.get(),
        throwsA(isA<MisarMailPlanLimitError>()
            .having((e) => e.plan, 'plan', 'starter')
            .having((e) => e.upgradeUrl, 'upgradeUrl',
                'https://misarmail.com/pricing')
            .having((e) => e.retryAfter, 'retryAfter', 600)
            .having((e) => e.feature, 'feature', 'emails')),
      );
      expect(attempts, equals(1),
          reason: 'a plan refusal must not burn the retry budget');
    });

    test('a plain rate-limit 429 is still retried', () async {
      var attempts = 0;
      final mock = MockClient((req) async {
        attempts++;
        if (attempts == 1) {
          return http.Response(jsonEncode({'error': 'Rate limit exceeded'}), 429,
              headers: {'content-type': 'application/json'});
        }
        return http.Response(
          jsonEncode({
            'plan': {'slug': 'starter', 'name': 'Starter'}
          }),
          200,
          headers: {'content-type': 'application/json'},
        );
      });

      final out = await _client(mock).plan.get();
      expect(attempts, equals(2));
      expect(out['plan']['slug'], equals('starter'));
    });
  });

  test('401 surfaces as MisarMailError', () async {
    final mock = MockClient((req) async => http.Response(
          jsonEncode({'error': 'Invalid or missing API key'}),
          401,
          headers: {'content-type': 'application/json'},
        ));

    await expectLater(
      _client(mock).plan.get(),
      throwsA(isA<MisarMailError>().having((e) => e.status, 'status', 401)),
    );
  });
}

File: go/client_test.go
package misarmail_test

import (
	"context"
	"errors"
	"net/http"
	"net/http/httptest"
	"testing"

	"github.com/Misar-AI/misarmail-sdks/go/v5/misarmail"
)

// newClient points the SDK at a test server.
func newClient(t *testing.T, h http.HandlerFunc) *misarmail.Client {
	t.Helper()
	srv := httptest.NewServer(h)
	t.Cleanup(srv.Close)
	return misarmail.New("mmk_test", misarmail.WithBaseURL(srv.URL))
}

func TestPlanGet(t *testing.T) {
	c := newClient(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/plan" {
			t.Errorf("path = %q, want /plan", r.URL.Path)
		}
		if got := r.Header.Get("Authorization"); got != "Bearer mmk_test" {
			t.Errorf("Authorization = %q", got)
		}
		w.Header().Set("Content-Type", "application/json")
		w.Write([]byte(`{
			"plan": {"slug":"starter","name":"Starter"},
			"sending": {"emails_per_day":500,"emails_per_month":10000},
			"usage": [{"feature":"emails","used":120,"limit":500,"remaining":380}],
			"upgrade": null
		}`))
	})

	got, err := c.Plan.Get(context.Background())
	if err != nil {
		t.Fatalf("Plan.Get: %v", err)
	}
	if got.Plan.Slug != "starter" {
		t.Errorf("plan slug = %q, want starter", got.Plan.Slug)
	}
	if len(got.Usage) != 1 || got.Usage[0].Used != 120 {
		t.Errorf("usage = %+v", got.Usage)
	}
	if got.Upgrade != nil {
		t.Errorf("upgrade should be nil while the plan has headroom")
	}
}

// A spent allowance answers 429 — identical by status to a rate limit — so the
// client must distinguish them by body and refuse to retry the plan refusal.
func TestPlanLimitIsNotRetried(t *testing.T) {
	attempts := 0
	c := newClient(t, func(w http.ResponseWriter, r *http.Request) {
		attempts++
		w.Header().Set("X-Misar-Plan", "starter")
		w.Header().Set("X-Misar-Upgrade-Url", "https://misarmail.com/pricing")
		w.Header().Set("Retry-After", "600")
		w.WriteHeader(http.StatusTooManyRequests)
		w.Write([]byte(`{
			"error": "Monthly send allowance spent.",
			"code": "plan_limit_exceeded",
			"upgrade": {"feature":"emails","currentPlanSlug":"starter"}
		}`))
	})

	_, err := c.Plan.Get(context.Background())

	var limit *misarmail.PlanLimitError
	if !errors.As(err, &limit) {
		t.Fatalf("error = %T (%v), want *PlanLimitError", err, err)
	}
	if attempts != 1 {
		t.Errorf("attempts = %d, want 1 — a plan refusal must not burn the retry budget", attempts)
	}
	if limit.Plan != "starter" {
		t.Errorf("Plan = %q, want starter", limit.Plan)
	}
	if limit.UpgradeURL != "https://misarmail.com/pricing" {
		t.Errorf("UpgradeURL = %q", limit.UpgradeURL)
	}
	if limit.RetryAfter != 600 {
		t.Errorf("RetryAfter = %d, want 600", limit.RetryAfter)
	}
	if limit.Feature != "emails" {
		t.Errorf("Feature = %q, want emails", limit.Feature)
	}
}

// A plain rate-limit 429 carries no plan marker and must still be retried.
func TestRateLimitIsStillRetried(t *testing.T) {
	attempts := 0
	c := newClient(t, func(w http.ResponseWriter, r *http.Request) {
		attempts++
		if attempts == 1 {
			w.WriteHeader(http.StatusTooManyRequests)
			w.Write([]byte(`{"error":"Rate limit exceeded"}`))
			return
		}
		w.Write([]byte(`{"plan":{"slug":"starter","name":"Starter"}}`))
	})

	if _, err := c.Plan.Get(context.Background()); err != nil {
		t.Fatalf("Plan.Get after retry: %v", err)
	}
	if attempts != 2 {
		t.Errorf("attempts = %d, want 2", attempts)
	}
}

func TestAPIErrorOnUnauthorized(t *testing.T) {
	c := newClient(t, func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusUnauthorized)
		w.Write([]byte(`{"error":"Invalid or missing API key"}`))
	})

	_, err := c.Plan.Get(context.Background())

	var apiErr *misarmail.APIError
	if !errors.As(err, &apiErr) {
		t.Fatalf("error = %T, want *APIError", err)
	}
	if apiErr.Status != http.StatusUnauthorized {
		t.Errorf("status = %d", apiErr.Status)
	}
}
